### PR TITLE
Fix Thread diagnostic nullable attributes so they read as null when …

### LINKED ISSUE #26908: GenericThreadStackManagerImpl_OpenThread has non-spec-compliant encoding of some fields
When doing a wildcard read of a Thread device implementing default ThreadDiagnostics cluster (GenericThreadStackManagerImpl_OpenThread), we get failures on reading/subscribing to several attributes in the cluster when not yet associated to a Thread mesh:

```
chip.clusters.Objects.ThreadNetworkDiagnostics.Attributes.ActiveTimestamp: ValueDecodeFailure(TLVValue=None, Reason=InteractionModelError(<Status.Failure: 1>)),
   chip.clusters.Objects.ThreadNetworkDiagnostics.Attributes.PendingTimestamp: ValueDecodeFailure(TLVValue=None, Reason=InteractionModelError(<Status.Failure: 1>)),
   chip.clusters.Objects.ThreadNetworkDiagnostics.Attributes.Delay: ValueDecodeFailure(TLVValue=None, Reason=InteractionModelError(<Status.Failure: 1>)),
   chip.clusters.Objects.ThreadNetworkDiagnostics.Attributes.SecurityPolicy: ValueDecodeFailure(TLVValue=None, Reason=InteractionModelError(<Status.Failure: 1>)),
   chip.clusters.Objects.ThreadNetworkDiagnostics.Attributes.ChannelPage0Mask: ValueDecodeFailure(TLVValue=None, Reason=InteractionModelError(<Status.Failure: 1>)),
   chip.clusters.Objects.ThreadNetworkDiagnostics.Attributes.OperationalDatasetComponents: ValueDecodeFailure(TLVValue=None, Reason=InteractionModelError(<Status.Failure: 1>)),
```

All the above are only valid when Thread is properly configured.
The code (simplified slightly for brevety) in `src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp` in `GenericThreadStackManagerImpl_OpenThread<ImplClass>::_WriteThreadNetworkDiagnosticAttributeToTlv` has, for example:

```
    case ThreadNetworkDiagnostics::Attributes::ActiveTimestamp::Id: {
        err = CHIP_ERROR_INCORRECT_STATE;
        if (otDatasetIsCommissioned(mOTInst))
        {
            otOperationalDataset activeDataset;
            otError otErr = otDatasetGetActive(mOTInst, &activeDataset);
            VerifyOrExit(otErr == OT_ERROR_NONE, err = MapOpenThreadError(otErr));
            uint64_t activeTimestamp  = activeDataset.mActiveTimestamp;
            err = encoder.Encode(activeTimestamp);
        }
    }
    break;
```

This will simply return a CHIP_ERROR (which will lead to generic IM error ("FAILURE") instead of writing a null like the spec *requires*:

> 11.13.6.57. ActiveTimestamp Attribute
>
> This attribute SHALL be null when there is no dataset configured.

It appears all the problematic FAILURE are of that nature.

### PR DESCRIPTION
**Pull request by jmartinez-silabs**
…the device is not associated with a thread network. ~~Add an encoding method to directly encode null when the provided condition evaluates to false~~

Update: we encoded null when no thread dataset has been provided to the device rather than the attachment state as valuable data can be fetched with the dataset for all those listed nullable attributes even if thread is detached or disabled

fixes #26908

### CODE REVIEW

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp [hunk 1]
```diff
@@ -1055,82 +1055,94 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_WriteThreadNetw
...
+            uint16_t panId = otLinkGetPanId(mOTInst);
+            return pAttrEncoder->Encode(panId);
+        });
     }
     break;
 
     case ThreadNetworkDiagnostics::Attributes::ExtendedPanId::Id: {
-        const otExtendedPanId * pExtendedPanid = otThreadGetExtendedPanId(mOTInst);
-        err                                    = encoder.Encode(Encoding::BigEndian::Get64(pExtendedPanid->m8));
+        err = encoder.EncodeNullableOnCondition(mIsAttached, [this](const auto pAttrEncoder) -> CHIP_ERROR {
```
Comment (tcarmelveilleux) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1129]: A much cheaper way would be to do a `if (!mIsAttached) { return EncodeNull(); }` or surch.

Comment (jmartinez-silabs) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1129]: I'll compare code size, I jsut wanted to avoid a bunch of addition if else in that use case. But sure, if there is a big size impact and can change a if in each case

File: src/app/AttributeAccessInterface.h [hunk 2]
```diff
@@ -200,6 +200,35 @@ class AttributeValueEncoder
...
+     *        This method should only be used to encode nullable attributes
+     *        aCallback is not called when the condition evaluates to false
+     *
+     * @tparam AttributeEncoder
+     * @param condition When true, use aCallback to encode the attribute data. When false, encode null
+     * @param aCallback Provide your lamda function encoding the data.
+     * @return CHIP_ERROR Encoding succeed or failed with error code
+     */
+    template <typename AttributeEncoder>
+    CHIP_ERROR EncodeNullableOnCondition(bool condition, AttributeEncoder aCallback)
```
Comment (bzbarsky-apple) [inline src/app/AttributeAccessInterface.h:220]: The naming is pretty opaque; nothing about it says "will encode null if condition is false".

Why is this significantly better than consumers either putting their data in a Nullable and then calling `Encode` on it, or just having consumers branch on the condition themselves?

If we just want a helper for the OpenThread code, that helper could be file-local....

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp [hunk 3]
```diff
@@ -1255,32 +1267,42 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_WriteThreadNetw
...
+            uint32_t partitionId = otThreadGetPartitionId(mOTInst);
+            return pAttrEncoder->Encode(partitionId);
+        });
     }
     break;
 
     case ThreadNetworkDiagnostics::Attributes::Weighting::Id: {
-        uint8_t weight = otThreadGetLeaderWeight(mOTInst);
-        err            = encoder.Encode(weight);
+        err = encoder.EncodeNullableOnCondition(mIsAttached, [this](const auto pAttrEncoder) -> CHIP_ERROR {
```
Comment (bzbarsky-apple) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1278]: Here's a question.  Are there _any_ things in this switch that don't depend on mIsAttached?  And if not, could we just check for it up front?

Comment (jmartinez-silabs) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1278]: A bunch of attributes are not nullable and with return counter values of the thread instance. So I would say no it it's not at large but still I'll do it differently with a starting. I'll do that tomorrow morning

File: src/app/AttributeAccessInterface.h (hunk 2 above)
Comment (jmartinez-silabs) [reply src/app/AttributeAccessInterface.h:220]: I did implementation formmy use case in mind but thought it could be useful for others..maybe other diagnostic clusters attributes. Any how, based on Tennessee's comment, I compared the code size difference and it is not worth it. I'll change the code tomorrow morning .

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp [hunk 4]
```diff
@@ -1055,82 +1055,94 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_WriteThreadNetw
     switch (attributeId)
     {
     case ThreadNetworkDiagnostics::Attributes::Channel::Id: {
-        uint16_t channel = otLinkGetChannel(mOTInst);
-        err              = encoder.Encode(channel);
+        err = encoder.EncodeNullableOnCondition(mIsAttached, [this](const auto pAttrEncoder) -> CHIP_ERROR {
```
Comment (Damian-Nordic) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1058]: Are you sure this should return null when the Thread device is not attached? The spec only says: "A value of null
SHALL indicate that the Thread interface is not currently configured or operational." and I would interpret that as "Thread role is 'disabled'". Technically, it shouldn't matter for most attributes as anyway, to read them the device must be attached or the device being commissioned over BLE, but I can imagine some commissioners might want to read `RoutingRole` over BLE to learn if the device is a REED or MED etc.

Comment (jmartinez-silabs) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1058]: >  Thread interface is not currently configured or operational.

 Doesn't Thread role being disable fall into the category "not operation"? To me when thread is not in an attached mode, it is not operational.

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp (hunk 3 above)
Comment (tcarmelveilleux) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1278]: Let's just make sure we don't null some attributes that don't need nulling. Some of the offending attributes in the original issue (e.g. ActiveTimestamp) require an active dataset.

Some others (e.g. channel) rely on being configured (active dataset) or "operational".

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp (hunk 1 above)
Comment (jmartinez-silabs) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp]: I applied what we discussed on slack

File: src/app/AttributeAccessInterface.h (hunk 2 above)
Comment (jmartinez-silabs) [reply src/app/AttributeAccessInterface.h]: I removed the helper entirely

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp (hunk 3 above)
Comment (jmartinez-silabs) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp]: So except for the RoutingRole that can always fetch a relevant state from the thread instance, all other nullable attributes depend on an active dataset provided to the device. If that dataset is commissioned, even if the thread radio is deactivated or not yet attached, relevant data can be fetched.

If the thread instance doesn't have a valid dataset, then all those attributes will read as null.


Bottom line, whatever the attachment state of the device (operational or not), if the relevant data is fetchable (and valid) return that, else return null.

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp [hunk 5]
```diff
@@ -1052,6 +1052,35 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_WriteThreadNetw
 
     namespace ThreadNetworkDiagnostics = app::Clusters::ThreadNetworkDiagnostics;
 
+    if (otDatasetIsCommissioned(mOTInst))
+    {
+        // For the following nullable attributes of the cluster, encodeNull since
+        // thread instance cannot provide the related data when it is not currently configured.
+        //
+        // Note that RoutingRole is nullable but not listed here as thread provide
```
Comment (bzbarsky-apple) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1060]:
```suggestion
        // Note that RoutingRole is nullable but not listed here as thread provides
```

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp [hunk 6]
```diff
@@ -1052,6 +1052,35 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_WriteThreadNetw
 
     namespace ThreadNetworkDiagnostics = app::Clusters::ThreadNetworkDiagnostics;
 
+    if (otDatasetIsCommissioned(mOTInst))
```
Comment (bzbarsky-apple) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1055]: Is this check backwards?
```suggestion
    if (!otDatasetIsCommissioned(mOTInst))
```

**Review by bzbarsky-apple — COMMENTED** [review]
Just to check: various of these attributes (Channel, NetworkName) did not use to check `otDatasetIsCommissioned`.  But they should have?

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp (hunk 5 above)
Comment (bzbarsky-apple) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1060]: Also, isn't RoutingRole determined dynamically in many cases, in which case it depends on not just having a dataset but in fact being attached?  I feel like this really needs to get clarified in the spec....

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp (hunk 6 above)
Comment (jmartinez-silabs) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1055]: Oh my, yes totally. I invert the condition to test the null encoded response case and forgot to switch it back when commiting

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp (hunk 5 above)
Comment (jmartinez-silabs) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1060]: RoutingRole also combines the state of the thread network. In a uncommissionned state, the device role/state would be detached, and when the radio is turned off it is disabled

The spec does define those two values in the RoutingRole enum, 

Disabled -> unspecified
Detached -> detached

Since they are valid values of the enum and maps to what the thread stack provides, I believe the nullable value isn't needed as then other values provide more information

**Comment by jmartinez-silabs** [comment]
@bzbarsky-apple 
> Just to check: various of these attributes (Channel, NetworkName) did not use to check `otDatasetIsCommissioned`.  But they should have?

They didn't do any check before, but when uncommissionnedz the thread stack returns some default values that does not mean much toward what the commissioned thread network will be.

For instance
Network name will return OpenThread,
Channel will return 11,
PanId return 0,
Etc.

My opinion is that in and uncommissioned state, null is a better representation for the user

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp (hunk 5 above)
Comment (bzbarsky-apple) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp]: Ah, I see.  Great!

**Review by bzbarsky-apple — APPROVED** [review]

**Review by Damian-Nordic — APPROVED** [review]

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp (hunk 4 above)
Comment (Damian-Nordic) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp]: Thanks, looking good now!

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp [hunk 7]
```diff
@@ -1537,125 +1566,99 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_WriteThreadNetw
...
-        err = CHIP_ERROR_INCORRECT_STATE;
-
-        if (otDatasetIsCommissioned(mOTInst))
-        {
-            otOperationalDataset activeDataset;
-            otError otErr = otDatasetGetActive(mOTInst, &activeDataset);
-            VerifyOrExit(otErr == OT_ERROR_NONE, err = MapOpenThreadError(otErr));
+        otOperationalDataset activeDataset;
+        otError otErr = otDatasetGetActive(mOTInst, &activeDataset);
+        VerifyOrReturnError(otErr == OT_ERROR_NONE, MapOpenThreadError(otErr));
```
Comment (tcarmelveilleux) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1608]: Is there a way to factor this code out of several switch cases? Can be done as a follow-up.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp (hunk 7 above)
Comment (jmartinez-silabs) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1608]: Ok I will think of something and I'll push something soon
